Miscellaneous __torch_function__ fixes

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75484

I figured these out by unconditionally turning on a no-op torch function
mode on the test suite and then fixing errors as they showed up.  Here's
what I found:

- _parse_to failed internal assert when __torch_function__'ed because it
  claims its name is "to" to the argument parser; added a name override
  so we know how to find the correct name

- Infix operator magic methods on Tensor did not uniformly handle
  __torch_function__ and TypeError to NotImplemented.  Now, we always
  do the __torch_function__ handling in
  _wrap_type_error_to_not_implemented and your implementation of
  __torch_function__ gets its TypeErrors converted to NotImplemented
  (for better or for worse; see
  https://github.com/pytorch/pytorch/issues/75462 )

- A few cases where code was incorrectly testing if a Tensor was
  Tensor-like in the wrong way, now use is_tensor_like (in grad
  and in distributions).  Also update docs for has_torch_function to
  push people to use is_tensor_like.

- is_grads_batched was dropped from grad in handle_torch_function, now
  fixed

- Report that you have a torch function even if torch function is
  disabled if a mode is enabled.  This makes it possible for a mode
  to return NotImplemented, pass to a subclass which does some
  processing and then pass back to the mode even after the subclass
  disables __torch_function__ (so the tensors are treated "as if"
  they are regular Tensors).  This brings the C++ handling behavior
  in line with the Python behavior.

- Make the Python implementation of overloaded types computation match
  the C++ version: when torch function is disabled, there are no
  overloaded types (because they all report they are not overloaded).

- torch.autograd.grad did not faithfully report its arguments to handle_torch_function; now it does

- Wrapper in test/test_overrides.py did not record usages for all arguments, leading to spurious failure when torch.autograd.grad started correctly wrapping Wrapper in a tuple (and not only true Tensors)

- tensordot did not forward out kwarg to handle_torch_function

- multi_head_attention_forward did not forward average_attn_weights kwarg to handle_torch_function

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Miscellaneous __torch_function__ fixes

I figured these out by unconditionally turning on a no-op torch function
mode on the test suite and then fixing errors as they showed up.  Here's
what I found:

- _parse_to failed internal assert when __torch_function__'ed because it
  claims its name is "to" to the argument parser; added a name override
  so we know how to find the correct name

- Infix operator magic methods on Tensor did not uniformly handle
  __torch_function__ and TypeError to NotImplemented.  Now, we always
  do the __torch_function__ handling in
  _wrap_type_error_to_not_implemented and your implementation of
  __torch_function__ gets its TypeErrors converted to NotImplemented
  (for better or for worse; see
  https://github.com/pytorch/pytorch/issues/75462 )

- A few cases where code was incorrectly testing if a Tensor was
  Tensor-like in the wrong way, now use is_tensor_like (in grad
  and in distributions).  Also update docs for has_torch_function to
  push people to use is_tensor_like.

- is_grads_batched was dropped from grad in handle_torch_function, now
  fixed

- Report that you have a torch function even if torch function is
  disabled if a mode is enabled.  This makes it possible for a mode
  to return NotImplemented, pass to a subclass which does some
  processing and then pass back to the mode even after the subclass
  disables __torch_function__ (so the tensors are treated "as if"
  they are regular Tensors).  This brings the C++ handling behavior
  in line with the Python behavior.

- Make the Python implementation of overloaded types computation match
  the C++ version: when torch function is disabled, there are no
  overloaded types (because they all report they are not overloaded).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

zou3519 commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -274,8 +274,7 @@ inline PythonArgs PythonArgParser::parse(PyObject* self, ParsedArgs<0>& dst) {
   return !this->signature.overloaded_args.empty() ||
-      (torch::torch_function_enabled() &&
-       at::impl::PythonTorchFunctionTLS::get_mode());
+       at::impl::PythonTorchFunctionTLS::get_mode();
```
hmmm, this is kind of weird now because now we're saying that torch function mode doesn't respect _disabled_torch_function_impl. Is this an argument for torch_function mode having a separate name from torch_function?

ezyang replied:
Yes, it is confusing. The intended use case for having these be separate is to allow both "mode runs first, then subclass" and "subclass runs first, then mode", where the subclass is an "is-a" style subclass (like the default Tensor subclass handling). In this world, you need to be able to disable subclass dispatch without also disabling mode dispatch (because you want mode to handle things next).

I'm not sure having the names being different would be that helpful; a more descriptive name for 'DisableTorchFunction' might help.

zou3519 reviewed (commented):
This mostly LGTM, I have some questions about the `_parse_to` fix and what exactly is happening in the infinite loop test case

peterbell10 commented on torch/_tensor.py:
```diff
@@ -659,6 +654,7 @@ def __format__(self, format_spec):
 
+    @_wrap_type_error_to_not_implemented
     def __ipow__(self, other):  # type: ignore[misc]
         if has_torch_function_variadic(self, other):
             return handle_torch_function(Tensor.__ipow__, (self, other), self, other)
```
```suggestion
```

ezyang commented on tools/autograd/templates/python_nn_functions.cpp:
```diff
@@ -43,7 +43,7 @@ static PyObject * THPVariable__parse_to(PyObject* module, PyObject* args, PyObje
   auto r = parser.parse(args, kwargs, parsed_args);
   if (r.has_torch_function()) {
-    return handle_torch_function(r, args, kwargs, THPNNVariableFunctionsModule, "torch.nn");
+    return handle_torch_function(r, args, kwargs, THPNNVariableFunctionsModule, "torch.nn", "_parse_to");
```
For (2), it's so the error message on invalid arguments refers to to (which is nn.Module.to, which makes sense to users). What I am not sure about is why this is `__torch_function__` overridable.

ezyang commented on torch/_tensor.py:
```diff
@@ -631,14 +632,10 @@ def unique_consecutive(self, return_inverse=False, return_counts=False, dim=None
 
     @_wrap_type_error_to_not_implemented
     def __rsub__(self, other):
-        if has_torch_function_variadic(self, other):
-            return handle_torch_function(Tensor.__rsub__, (self, other), self, other)
```
Yep

ezyang commented on test/test_overrides.py:
```diff
@@ -1210,6 +1242,88 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
+
+        self.assertTrue(called)
+
+    def test_mode_notimplemented_loop(self):
```
Previously, the Python codepath for `handle_torch_function` did not remove types from overloaded arguments when torch function was disabled. The infinite loop looked like this:

```
  File "/data/users/ezyang/pytorch-tmp/torch/_tensor.py", line 1122, in __torch_function__     ret = func(*args, **kwargs)
  File "/data/users/ezyang/pytorch-tmp/test/test_overrides.py", line 43, in bar
    return handle_torch_function(bar, (a,), a)                                
  File "/data/users/ezyang/pytorch-tmp/torch/overrides.py", line 1414, in handle_torch_func
tion                                                                                       
    result = torch_func_method(public_api, types, args, kwargs)
  File "/data/users/ezyang/pytorch-tmp/torch/_tensor.py", line 1122, in __torch_function__
    ret = func(*args, **kwargs)   
  File "/data/users/ezyang/pytorch-tmp/test/test_overrides.py", line 43, in bar            
    return handle_torch_function(bar, (a,), a)
  File "/data/users/ezyang/pytorch-tmp/torch/overrides.py", line 1414, in handle_torch_func
tion                                                                                       
    result = torch_func_method(public_api, types, args, kwargs)   
```

What's going on is that the mode is enabled (so `has_torch_function` reports true) but torch function is disabled; however, when we collect up overloaded types we don't respect torch function being disabled and attempt to call into the `Tensor.__torch_function__` implementation again, ergo infinite loop.

zou3519 replied:
Ahh okay, thanks for the clarification

ezyang committed:
Update on "Miscellaneous __torch_function__ fixes"


I figured these out by unconditionally turning on a no-op torch function
mode on the test suite and then fixing errors as they showed up.  Here's
what I found:

- _parse_to failed internal assert when __torch_function__'ed because it
  claims its name is "to" to the argument parser; added a name override
  so we know how to find the correct name

- Infix operator magic methods on Tensor did not uniformly handle
  __torch_function__ and TypeError to NotImplemented.  Now, we always
  do the __torch_function__ handling in
  _wrap_type_error_to_not_implemented and your implementation of
  __torch_function__ gets its TypeErrors converted to NotImplemented
  (for better or for worse; see
  https://github.com/pytorch/pytorch/issues/75462 )

- A few cases where code was incorrectly testing if a Tensor was
  Tensor-like in the wrong way, now use is_tensor_like (in grad
  and in distributions).  Also update docs for has_torch_function to
  push people to use is_tensor_like.

- is_grads_batched was dropped from grad in handle_torch_function, now
  fixed

- Report that you have a torch function even if torch function is
  disabled if a mode is enabled.  This makes it possible for a mode
  to return NotImplemented, pass to a subclass which does some
  processing and then pass back to the mode even after the subclass
  disables __torch_function__ (so the tensors are treated "as if"
  they are regular Tensors).  This brings the C++ handling behavior
  in line with the Python behavior.

- Make the Python implementation of overloaded types computation match
  the C++ version: when torch function is disabled, there are no
  overloaded types (because they all report they are not overloaded).

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Miscellaneous __torch_function__ fixes"


I figured these out by unconditionally turning on a no-op torch function
mode on the test suite and then fixing errors as they showed up.  Here's
what I found:

- _parse_to failed internal assert when __torch_function__'ed because it
  claims its name is "to" to the argument parser; added a name override
  so we know how to find the correct name

- Infix operator magic methods on Tensor did not uniformly handle
  __torch_function__ and TypeError to NotImplemented.  Now, we always
  do the __torch_function__ handling in
  _wrap_type_error_to_not_implemented and your implementation of
  __torch_function__ gets its TypeErrors converted to NotImplemented
  (for better or for worse; see
  https://github.com/pytorch/pytorch/issues/75462 )

- A few cases where code was incorrectly testing if a Tensor was
  Tensor-like in the wrong way, now use is_tensor_like (in grad
  and in distributions).  Also update docs for has_torch_function to
  push people to use is_tensor_like.

- is_grads_batched was dropped from grad in handle_torch_function, now
  fixed

- Report that you have a torch function even if torch function is
  disabled if a mode is enabled.  This makes it possible for a mode
  to return NotImplemented, pass to a subclass which does some
  processing and then pass back to the mode even after the subclass
  disables __torch_function__ (so the tensors are treated "as if"
  they are regular Tensors).  This brings the C++ handling behavior
  in line with the Python behavior.

- Make the Python implementation of overloaded types computation match
  the C++ version: when torch function is disabled, there are no
  overloaded types (because they all report they are not overloaded).

- torch.autograd.grad did not faithfully report its arguments to handle_torch_function; now it does

- Wrapper in test/test_overrides.py did not record usages for all arguments, leading to spurious failure when torch.autograd.grad started correctly wrapping Wrapper in a tuple (and not only true Tensors)

- tensordot did not forward out kwarg to handle_torch_function

- multi_head_attention_forward did not forward average_attn_weights kwarg to handle_torch_function

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Miscellaneous __torch_function__ fixes"


I figured these out by unconditionally turning on a no-op torch function
mode on the test suite and then fixing errors as they showed up.  Here's
what I found:

- _parse_to failed internal assert when __torch_function__'ed because it
  claims its name is "to" to the argument parser; added a name override
  so we know how to find the correct name

- Infix operator magic methods on Tensor did not uniformly handle
  __torch_function__ and TypeError to NotImplemented.  Now, we always
  do the __torch_function__ handling in
  _wrap_type_error_to_not_implemented and your implementation of
  __torch_function__ gets its TypeErrors converted to NotImplemented
  (for better or for worse; see
  https://github.com/pytorch/pytorch/issues/75462 )

- A few cases where code was incorrectly testing if a Tensor was
  Tensor-like in the wrong way, now use is_tensor_like (in grad
  and in distributions).  Also update docs for has_torch_function to
  push people to use is_tensor_like.

- is_grads_batched was dropped from grad in handle_torch_function, now
  fixed

- Report that you have a torch function even if torch function is
  disabled if a mode is enabled.  This makes it possible for a mode
  to return NotImplemented, pass to a subclass which does some
  processing and then pass back to the mode even after the subclass
  disables __torch_function__ (so the tensors are treated "as if"
  they are regular Tensors).  This brings the C++ handling behavior
  in line with the Python behavior.

- Make the Python implementation of overloaded types computation match
  the C++ version: when torch function is disabled, there are no
  overloaded types (because they all report they are not overloaded).

- torch.autograd.grad did not faithfully report its arguments to handle_torch_function; now it does

- Wrapper in test/test_overrides.py did not record usages for all arguments, leading to spurious failure when torch.autograd.grad started correctly wrapping Wrapper in a tuple (and not only true Tensors)

- tensordot did not forward out kwarg to handle_torch_function

- multi_head_attention_forward did not forward average_attn_weights kwarg to handle_torch_function

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

zou3519 reviewed (approved):
I'm really not sure about the `_parse_to` but everything else LGTM

ezyang commented:
@pytorchbot merge this
